features/steps/steps.go

```diff
@@ -2,87 +2,204 @@ package steps
-		return err
-	}
+	return nil
+}
 
-	signals := registerInterrupt()
+// theCallToAddInstanceDimensionIsSuccessful generates a mocked response for Dataset API
+// POST /instances/{id}/dimensions
+func (c *Component) theCallToAddInstanceDimensionIsSuccessful(id string) error {
+	c.DatasetAPI.NewHandler().
+		Post("/instances/"+id+"/dimensions").
+		Reply(http.StatusOK).
+		AddHeader("ETag", testETag)
 
-	// run application in separate goroutine
-	go func() {
-		c.svc.Start(context.Background(), c.errorChan)
-	}()
+	return nil
+}
 
-	// consume extracted observations
-	for _, e := range observationEvents {
-		if err := c.sendToConsumer(e); err != nil {
-			return err
-		}
-	}
+// theCallToUpdateInstanceIsSuccessful generates a mocked response for Dataset API
+// PUT /instances/{id}
+func (c *Component) theCallToUpdateInstanceIsSuccessful(id string) error {
+	c.DatasetAPI.NewHandler().
+		Put("/instances/"+id).
+		Reply(http.StatusOK).
+		AddHeader("ETag", testETag)
 
-	time.Sleep(300 * time.Millisecond)
+	return nil
+}
 
-	// kill application
-	signals <- os.Interrupt
+// theCallToUpdateJobIsSuccessful generates a mocked response for Import API
+// PUT /jobs/{id}
+func (c *Component) theCallToUpdateJobIsSuccessful(id string) error {
+	c.ImportAPI.NewHandler().
+		Put("/jobs/" + id).
+		Reply(http.StatusOK)
 
 	return nil
 }
 
-func (c *Component) convertToHelloEvents(table *godog.Table) ([]*event.CategoryDimensionImport, error) {
-	assist := assistdog.NewDefault()
-	events, err := assist.CreateSlice(&event.CategoryDimensionImport{}, table)
+// theseDimensionCompleteEventsAreProduced consumes kafka messages that are expected to be produced by the service under test
+// and validates that they match the expected values in the test
+func (c *Component) theseDimensionCompleteEventsAreProduced(events *godog.Table) error {
+	expected, err := assistdog.NewDefault().CreateSlice(new(event.InstanceComplete), events)
 	if err != nil {
-		return nil, err
+		return fmt.Errorf("failed to create slice from godog table: %w", err)
 	}
-	return events.([]*event.CategoryDimensionImport), nil
-}
 
-func (c *Component) sendToConsumer(e *event.CategoryDimensionImport) error {
-	bytes, err := schema.CategoryDimensionImport.Marshal(e)
-	if err != nil {
-		return err
+	var got []*event.InstanceComplete
+	listen := true
+
+	for listen {
+		select {
+		case <-time.After(DimensionCompleteEventsTimeout):
+			listen = false
+		case <-c.consumer.Channels().Closer:
+			return errors.New("closer channel closed")
+		case msg, ok := <-c.consumer.Channels().Upstream:
+			if !ok {
+				return errors.New("upstream channel closed")
+			}
+
+			var e event.InstanceComplete
+			var s = schema.InstanceComplete
+
+			if err := s.Unmarshal(msg.GetData(), &e); err != nil {
+				msg.Commit()
+				msg.Release()
+				return fmt.Errorf("error unmarshalling message: %w", err)
+			}
+
+			msg.Commit()
+			msg.Release()
+
+			got = append(got, &e)
+		}
+	}
+
+	if !reflect.DeepEqual(got, expected) {
```

Comment:
I also realised that printing here as it was wasn't good enough as you only got the addresses, however using `github.com/google/go-cmp/cmp` here may be a better option (which is what I went for), the comparison ends up a one liner and does everything for us but all type safe etc
```
	if diff := cmp.Diff(got, expected); diff != "" {
		return fmt.Errorf("-got +expected)\n%s\n", diff)
	}
```

Reply:
ok, I will change it. I added this method to do it:
```
func printValues(in []*event.InstanceComplete) string {
	ret := "[ "
	for _, val := range in {
		ret += fmt.Sprintf("%+v ", val)
	}
	return ret + "]"
}
```
But maybe it's better to use go-cmp